### DISCUSSION TITLE
What's the correct way to implement a TerminalReporter

### DISCUSSION THREAD

**Discussion by renanivo** [pytest-dev/pytest#10334] [Q&A] [answered]
Hi

Some time ago I tried to add mypy static type checking to [my plugin](https://github.com/renanivo/pytest-testdox). I've got an error because I have written a class which inherits from `TerminalReporter` (which is a `@final` class).

I've ssen that [it was intended](https://github.com/pytest-dev/pytest/pull/7780), but I don't know how can I have a custom reporter (or custom pytest output) without doing it. I'm looking at other plugins but I've found none that don't have the same issue.

So, is there a correct way to implement a `TerminalReporter` or this kind of plugin is expected to get this error?

**Comment by RonnyPfannschmidt** [comment, marked as answer]
I believe we need to remove the final marker as the terminalwriter plugin currently is a bit of a painful god object where reimplementation seems daunting

Cc @bluetech

**Comment by renanivo** [comment]
@RonnyPfannschmidt , @bluetech  Do you think the `@final` marker could be dropped from `TerminalReporter` in the future? Does that mean it could be available for importing from the `pytest` package? (We have to import `TerminalReporter` from `_pytest.terminal` nowadays, which means it is an internal class and should not be used elsewhere AFAIK).

**Reply by RonnyPfannschmidt** [reply]
it is a internal class that we dont want to expose as public api as its a legacy mess that needs replacement

**Reply by renanivo** [reply]
Got it. So probably the best thing I can do right now is to add `# type: ignore` to it (at my own risk) and wait for a public interface done with extension in mind.

Let me know if there's something I can do to help. I would love to contribute on a replacement, but I know probably there's a lot of legacy behavior that should be supported.

**Reply by RonnyPfannschmidt** [reply]
so far i have not digged into whats needed to safely replace it

topics ot take a look are replacing markup,, sorting out statusline/ux, having a more abstract ux interaction apis (so we can stream events into a textual ui in a way that works with xdist as well)

i'd like to enable general decoupling of the pytest core from "terminals" - ides, language servers and higher level uis are nice to have and blocked on foundation work since bascially decades
